google/gax/api_callable.py

```diff
@@ -382,6 +382,28 @@ def inner(*args, **kwargs):
     return inner
 
 
+def _merge_options_metadata(options, settings):
```

Comment:
That logic merges kwargs dictionary, just overwriting values, which share the same key. For this particular case we need to merge a specific element in that dictionary: kwargs['metadata']. It is like "deep merging" of a particular element in the dictionary. The settings.merge() performs only the "shallow merging".

That list (returned from kwargs['metadata']) consists of tuples, and merging here is performed using first element of a tuple as a key (one tuple can overwrite another if they their first elements are equal).